sdk/inc/azure/iot/az_iot_pnp_client.h

```diff
@@ -0,0 +1,670 @@
+    az_span component_name);
+
+/**
+ * @brief Read the IoT Plug and Play twin properties version for a given component
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] json_reader The #az_json_reader to parse through.
+ * @param[in] is_partial The boolean representing whether the twin document is from a partial update
+ * (#AZ_IOT_PNP_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES) or a full twin document
+ * (#AZ_IOT_PNP_CLIENT_TWIN_RESPONSE_TYPE_GET).
+ * @param[out] out_version The version of the properties in the json payload.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p json_reader must not be `NULL`.
+ * @pre \p out_version must not be `NULL`. It must point to a valid int32_t variable.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK If the function returned a valid version.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_twin_get_property_version(
+    az_iot_pnp_client const* client,
+    az_json_reader* json_reader,
+    bool is_partial,
+    int32_t* out_version);
+
+/**
+ * @brief Read the IoT Plug and Play twin properties component-by-component.
+ *
+ * Note that between calls, the #az_span pointed to by \p out_component_name shall not be modified,
+ * only checked and compared. Internally, the #az_span is only changed if the component name changes
+ * in the JSON document and is not necessarily set every invocation of the function.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] json_reader The #az_json_reader to parse through.
+ * @param[in] is_partial The boolean representing whether the twin document is from a partial update
+ * (#AZ_IOT_PNP_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES) or a full twin document
+ * (#AZ_IOT_PNP_CLIENT_TWIN_RESPONSE_TYPE_GET).
+ * @param[out] out_component_name The #az_span* representing the value of the component.
+ * @param[out] out_property_name The #az_json_token* representing the name of the property.
+ * @param[out] out_property_value The #az_json_reader* representing the value of the property.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p json_reader must not be `NULL`.
+ * @pre \p out_component_name must not be `NULL`. It must point to a valid #az_span instance.
+ * @pre \p out_property_name must not be `NULL`. It must point to a valid #az_json_token instance.
+ * @pre \p out_property_value must not be `NULL`. It must point to a valid #az_json_reader instance.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ * @retval #AZ_OK If the function returned a valid property name and value.
+ * @retval #AZ_IOT_END_OF_PROPERTIES If there are no more properties left for the component.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_twin_get_next_component_property(
+    az_iot_pnp_client const* client,
+    az_json_reader* json_reader,
+    bool is_partial,
+    az_span* out_component_name,
+    az_json_token* out_property_name,
```

Comment:
Is this memory need to be provided by caller? If yes, can we not make json_reader as cursor out value pointing to property name?

Reply:
Do you mean an output `az_json_reader` instead of `az_json_token`?